net/portmapper/upnp.go

```diff
@@ -0,0 +1,261 @@
+func addAnyPortMapping(
+	ctx context.Context,
+	upnp upnpClient,
+	externalPort uint16,
+	internalPort uint16,
+	internalClient string,
+	leaseDuration time.Duration,
+) (newPort uint16, err error) {
+	if upnp, ok := upnp.(*internetgateway2.WANIPConnection2); ok {
+		return upnp.AddAnyPortMapping(
+			ctx,
+			"",
+			externalPort,
+			"udp",
+			internalPort,
+			internalClient,
+			true,
+			tsPortMappingDesc,
+			uint32(leaseDuration.Seconds()),
+		)
+	}
+	err = upnp.AddPortMapping(
+		ctx,
+		"",
+		externalPort,
+		"udp",
+		internalPort,
+		internalClient,
+		true,
+		tsPortMappingDesc,
+		uint32(leaseDuration.Seconds()),
+	)
+	return internalPort, err
+}
+
+// getUPnPClients gets a client for interfacing with UPnP, ignoring the underlying protocol for
+// now.
+// Adapted from https://github.com/huin/goupnp/blob/master/GUIDE.md.
+func getUPnPClient(ctx context.Context, gw netaddr.IP) (client upnpClient, err error) {
+	ctx, cancel := context.WithTimeout(ctx, 250*time.Millisecond)
+	defer cancel()
+	tasks, _ := errgroup.WithContext(ctx)
+	// Attempt to connect over the multiple available connection types concurrently,
+	// returning the fastest.
+
+	// XXX: this url seems super brittle? maybe discovery is better but this is faster
+	u, err := url.Parse(fmt.Sprintf("http://%s:5000/rootDesc.xml", gw))
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO need to add a context to each of these function calls
+	var ip1Clients []*internetgateway2.WANIPConnection1
+	var errs [3]error
+	tasks.Go(func() error {
+		var err error
+		ip1Clients, err = internetgateway2.NewWANIPConnection1ClientsByURL(ctx, u)
+		errs[0] = err
+		return nil
+	})
+	var ip2Clients []*internetgateway2.WANIPConnection2
+	tasks.Go(func() error {
+		var err error
+		ip2Clients, err = internetgateway2.NewWANIPConnection2ClientsByURL(ctx, u)
+		errs[1] = err
+		return nil
+	})
+	var ppp1Clients []*internetgateway2.WANPPPConnection1
+	tasks.Go(func() error {
+		var err error
+		ppp1Clients, err = internetgateway2.NewWANPPPConnection1ClientsByURL(ctx, u)
+		errs[2] = err
+		return nil
+	})
+
+	err = tasks.Wait()
+
+	switch {
+	case len(ip2Clients) > 0:
+		return ip2Clients[0], nil
+	case len(ip1Clients) > 0:
+		return ip1Clients[0], nil
+	case len(ppp1Clients) > 0:
+		return ppp1Clients[0], nil
+	default:
+		if err != nil {
+			return nil, err
+		}
+		for i := range errs {
+			if errs[i] != nil {
+				err = errs[i]
+				break
+			}
+		}
+		// Didn't get any outputs, report if there was an error or nil if
+		// just no clients.
+		return nil, err
+	}
+}
+
+// getUPnPPortMapping attempts to create a port-mapping over the UPnP protocol. On success,
+// it will return the externally exposed IP and port. Otherwise, it will return a zeroed IP and
+// port and an error.
+func (c *Client) getUPnPPortMapping(
+	ctx context.Context,
+	gw netaddr.IP,
+	internal netaddr.IPPort,
+	prevPort uint16,
+) (external netaddr.IPPort, err error) {
```

Comment:
maybe return an `ok bool` instead of err if all your errors are going to be `NoMappingError{ErrNoPortMappingServices}` ?  That error would look better at the call site probably anyway.